File: iOS/KinKitchen/KinKitchen/Services/Supabase/ProfileService.swift
//
//  ProfileService.swift
//  KinKitchen
//
//  Created by Greg Hudler on 8/27/26.
//

import Foundation
import Supabase

enum UsernameAvailability {
    case unknown
    case checking
    case available
    case taken
}

// Any additional requirements are added here and
// in the missingProfileRequirements function below.
enum ProfileRequirement: CaseIterable {
    case firstName
    case lastName
    case username
    case birthDate
}

enum ProfileService {

    static func fetchCurrentProfile() async throws -> Profile {

        let user =
            try await SupabaseManager.client.auth.session.user

        let profile: Profile =
            try await SupabaseManager.client
                .from("profiles")
                .select()
                .eq("id", value: user.id)
                .single()
                .execute()
                .value

        return profile
    }

    static func isUsernameAvailable(
        _ username: String
    ) async throws -> Bool {

        let cleanUsername =
            username.trimmingCharacters(
                in: .whitespacesAndNewlines
            )

        guard !cleanUsername.isEmpty else {
            return false
        }

        let existingProfiles: [Profile] =
            try await SupabaseManager.client
                .from("profiles")
                .select()
                .eq(
                    "username",
                    value: cleanUsername
                )
                .limit(1)
                .execute()
                .value

        return existingProfiles.isEmpty
    }

    static func isCurrentProfileComplete() async throws -> Bool {

        let profile =
            try await fetchCurrentProfile()

        return missingProfileRequirements(
            for: profile
        ).isEmpty
    }

    static func completeInitialProfile(
        firstName: String,
        lastName: String,
        username: String,
        location: String,
        birthDate: Date,
        bio: String
    ) async throws {

        let user =
            try await SupabaseManager.client.auth.session.user

        let cleanFirstName =
            firstName.trimmingCharacters(
                in: .whitespacesAndNewlines
            )

        let cleanLastName =
            lastName.trimmingCharacters(
                in: .whitespacesAndNewlines
            )

        let displayName =
            "\(cleanFirstName) \(cleanLastName)"

        let updates =
            InitialProfileUpdate(
                firstName: cleanFirstName,
                lastName: cleanLastName,
                displayName: displayName,
                username:
                    username.trimmingCharacters(
                        in: .whitespacesAndNewlines
                    ),
                location:
                    location.trimmingCharacters(
                        in: .whitespacesAndNewlines
                    ),
                birthDate:
                    dateString(
                        from: birthDate
                    ),
                bio:
                    bio.trimmingCharacters(
                        in: .whitespacesAndNewlines
                    )
            )

        try await SupabaseManager.client
            .from("profiles")
            .update(updates)
            .eq("id", value: user.id)
            .execute()
    }

    static func updateProfile(
        firstName: String,
        lastName: String,
        displayName: String,
        username: String,
        location: String,
        birthDate: Date,
        bio: String
    ) async throws {

        let user =
            try await SupabaseManager.client.auth.session.user

        let updates =
            ProfileUpdate(
                firstName:
                    firstName.trimmingCharacters(
                        in: .whitespacesAndNewlines
                    ),
                lastName:
                    lastName.trimmingCharacters(
                        in: .whitespacesAndNewlines
                    ),
                displayName:
                    displayName.trimmingCharacters(
                        in: .whitespacesAndNewlines
                    ),
                username:
                    username.trimmingCharacters(
                        in: .whitespacesAndNewlines
                    ),
                location:
                    location.trimmingCharacters(
                        in: .whitespacesAndNewlines
                    ),
                birthDate:
                    dateString(
                        from: birthDate
                    ),
                bio:
                    bio.trimmingCharacters(
                        in: .whitespacesAndNewlines
                    )
            )

        try await SupabaseManager.client
            .from("profiles")
            .update(updates)
            .eq("id", value: user.id)
            .execute()
    }

    static func uploadProfilePhoto(
        _ imageData: Data
    ) async throws -> String {

        let user =
            try await SupabaseManager.client.auth.session.user

        let path =
            "\(user.id.uuidString.lowercased())/profile.jpg"

        try await SupabaseManager.client.storage
            .from("profile-photos")
            .upload(
                path,
                data: imageData,
                options: FileOptions(
                    contentType: "image/jpeg",
                    upsert: true
                )
            )

        let updates =
            ProfilePhotoUpdate(
                profilePhotoPath: path
            )

        try await SupabaseManager.client
            .from("profiles")
            .update(updates)
            .eq("id", value: user.id)
            .execute()

        return path
    }

    static func fetchProfilePhoto(
        path: String
    ) async throws -> Data {

        try await SupabaseManager.client.storage
            .from("profile-photos")
            .download(
                path: path
            )
    }

    // MARK: - Dietary Setup

    static func isDietarySetupComplete() async throws -> Bool {

        let status =
            try await fetchDietarySetupStatus()

        return status.dietarySetupCompletedAt != nil
    }

    static func isDietaryReviewDue() async throws -> Bool {

        let status =
            try await fetchDietarySetupStatus()

        guard let completedAt =
            status.dietarySetupCompletedAt
        else {
            return false
        }

        let calendar =
            Calendar.current

        guard let reviewDate =
            calendar.date(
                byAdding: .year,
                value: 1,
                to: completedAt
            )
        else {
            return false
        }

        return Date() >= reviewDate
    }
    
    static func fetchDietaryLastUpdated() async throws -> Date? {

        let status =
            try await fetchDietarySetupStatus()

        return status.dietarySetupCompletedAt
    }

    static func markDietarySetupReviewed() async throws {

        let user =
            try await SupabaseManager.client.auth.session.user

        let updates =
            DietarySetupReviewUpdate(
                dietarySetupCompletedAt: Date()
            )

        try await SupabaseManager.client
            .from("profiles")
            .update(updates)
            .eq("id", value: user.id)
            .execute()
    }

    private static func fetchDietarySetupStatus() async throws
        -> DietarySetupStatus {

        let user =
            try await SupabaseManager.client.auth.session.user

        let status: DietarySetupStatus =
            try await SupabaseManager.client
                .from("profiles")
                .select("dietary_setup_completed_at")
                .eq("id", value: user.id)
                .single()
                .execute()
                .value

        return status
    }

    // MARK: - Dates

    static func date(
        from string: String?
    ) -> Date? {

        guard let string else {
            return nil
        }

        return birthDateFormatter.date(
            from: string
        )
    }

    private static func dateString(
        from date: Date
    ) -> String {

        birthDateFormatter.string(
            from: date
        )
    }

    private static let birthDateFormatter: DateFormatter = {

        let formatter =
            DateFormatter()

        formatter.calendar =
            Calendar(
                identifier: .gregorian
            )

        formatter.locale =
            Locale(
                identifier: "en_US_POSIX"
            )

        formatter.dateFormat =
            "yyyy-MM-dd"

        return formatter
    }()

    // Other requirements are added here.
    static func missingProfileRequirements(
        for profile: Profile
    ) -> [ProfileRequirement] {

        var missing:
            [ProfileRequirement] = []

        let firstName =
            profile.firstName?
                .trimmingCharacters(
                    in: .whitespacesAndNewlines
                )
                ?? ""

        let lastName =
            profile.lastName?
                .trimmingCharacters(
                    in: .whitespacesAndNewlines
                )
                ?? ""

        let username =
            profile.username?
                .trimmingCharacters(
                    in: .whitespacesAndNewlines
                )
                ?? ""

        let birthDate =
            profile.birthDate?
                .trimmingCharacters(
                    in: .whitespacesAndNewlines
                )
                ?? ""

        if firstName.isEmpty {
            missing.append(.firstName)
        }

        if lastName.isEmpty {
            missing.append(.lastName)
        }

        if username.isEmpty {
            missing.append(.username)
        }

        if birthDate.isEmpty {
            missing.append(.birthDate)
        }

        return missing
    }
}

private struct InitialProfileUpdate: Encodable {

    let firstName: String
    let lastName: String
    let displayName: String
    let username: String
    let location: String
    let birthDate: String
    let bio: String

    enum CodingKeys: String, CodingKey {

        case firstName =
            "first_name"

        case lastName =
            "last_name"

        case displayName =
            "display_name"

        case username
        case location

        case birthDate =
            "birth_date"

        case bio
    }
}

private struct ProfileUpdate: Encodable {

    let firstName: String
    let lastName: String
    let displayName: String
    let username: String
    let location: String
    let birthDate: String
    let bio: String

    enum CodingKeys: String, CodingKey {

        case firstName =
            "first_name"

        case lastName =
            "last_name"

        case displayName =
            "display_name"

        case username
        case location

        case birthDate =
            "birth_date"

        case bio
    }
}

private struct ProfilePhotoUpdate: Encodable {

    let profilePhotoPath: String

    enum CodingKeys: String, CodingKey {

        case profilePhotoPath =
            "profile_photo_path"
    }
}

private struct DietarySetupStatus: Decodable {

    let dietarySetupCompletedAt: Date?

    enum CodingKeys: String, CodingKey {

        case dietarySetupCompletedAt =
            "dietary_setup_completed_at"
    }
}

private struct DietarySetupReviewUpdate: Encodable {

    let dietarySetupCompletedAt: Date

    enum CodingKeys: String, CodingKey {

        case dietarySetupCompletedAt =
            "dietary_setup_completed_at"
    }
}

File: Android/app/src/main/java/com/pushtomaindev/kinkitchen/services/supabase/ProfileService.kt
package com.pushtomaindev.kinkitchen.services.supabase

import com.pushtomaindev.kinkitchen.models.user.Profile
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.storage.storage
import kotlinx.datetime.Clock
import kotlinx.datetime.DateTimeUnit
import kotlinx.datetime.TimeZone
import kotlinx.datetime.plus
import kotlinx.datetime.Instant
import kotlinx.datetime.LocalDate
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Mirrors iOS `UsernameAvailability`. */
enum class UsernameAvailability { UNKNOWN, CHECKING, AVAILABLE, TAKEN }

/**
 * Mirrors iOS `ProfileRequirement`. Any new requirement is added here and in
 * [ProfileService.missingProfileRequirements].
 */
enum class ProfileRequirement { FIRST_NAME, LAST_NAME, USERNAME, BIRTH_DATE }

/** Mirrors iOS `ProfileService`. */
object ProfileService {

    private const val PHOTO_BUCKET = "profile-photos"

    private val client get() = SupabaseManager.client

    // MARK: - Profile

    suspend fun fetchCurrentProfile(): Profile {
        val userId = requireUserId()
        return client.from("profiles")
            .select { filter { eq("id", userId) } }
            .decodeSingle()
    }

    suspend fun isUsernameAvailable(username: String): Boolean {
        val clean = username.trim()
        if (clean.isEmpty()) return false

        return client.from("profiles")
            .select { filter { eq("username", clean) }; limit(1) }
            .decodeList<Profile>()
            .isEmpty()
    }

    suspend fun isCurrentProfileComplete(): Boolean =
        missingProfileRequirements(fetchCurrentProfile()).isEmpty()

    suspend fun completeInitialProfile(
        firstName: String,
        lastName: String,
        username: String,
        location: String,
        birthDate: LocalDate,
        bio: String,
    ) {
        val userId = requireUserId()
        val cleanFirst = firstName.trim()
        val cleanLast = lastName.trim()

        val updates = ProfileUpdate(
            firstName = cleanFirst,
            lastName = cleanLast,
            displayName = "$cleanFirst $cleanLast",
            username = username.trim(),
            location = location.trim(),
            birthDate = birthDate.toString(),
            bio = bio.trim(),
        )

        client.from("profiles").update(updates) { filter { eq("id", userId) } }
    }

    suspend fun updateProfile(
        firstName: String,
        lastName: String,
        displayName: String,
        username: String,
        location: String,
        birthDate: LocalDate,
        bio: String,
    ) {
        val userId = requireUserId()

        val updates = ProfileUpdate(
            firstName = firstName.trim(),
            lastName = lastName.trim(),
            displayName = displayName.trim(),
            username = username.trim(),
            location = location.trim(),
            birthDate = birthDate.toString(),
            bio = bio.trim(),
        )

        client.from("profiles").update(updates) { filter { eq("id", userId) } }
    }

    // MARK: - Photos

    suspend fun uploadProfilePhoto(imageData: ByteArray): String {
        val userId = requireUserId()
        val path = "${userId.lowercase()}/profile.jpg"

        client.storage.from(PHOTO_BUCKET).upload(path, imageData) {
            upsert = true
            contentType = io.ktor.http.ContentType.Image.JPEG
        }

        client.from("profiles")
            .update(ProfilePhotoUpdate(path)) { filter { eq("id", userId) } }

        return path
    }

    suspend fun fetchProfilePhoto(path: String): ByteArray =
        client.storage.from(PHOTO_BUCKET).downloadAuthenticated(path)

    // MARK: - Dietary Setup

    suspend fun isDietarySetupComplete(): Boolean =
        fetchDietarySetupStatus().dietarySetupCompletedAt != null

    /** Mirrors iOS `isDietaryReviewDue` — a review falls due one year on. */
    suspend fun isDietaryReviewDue(): Boolean {
        val completedAt = fetchDietarySetupStatus().dietarySetupCompletedAt ?: return false
        val reviewDate = completedAt.plus(365, DateTimeUnit.DAY, TimeZone.currentSystemDefault())
        return Clock.System.now() >= reviewDate
    }

    suspend fun fetchDietaryLastUpdated(): Instant? =
        fetchDietarySetupStatus().dietarySetupCompletedAt

    suspend fun markDietarySetupReviewed() {
        val userId = requireUserId()
        client.from("profiles")
            .update(DietarySetupReviewUpdate(Clock.System.now())) { filter { eq("id", userId) } }
    }

    private suspend fun fetchDietarySetupStatus(): DietarySetupStatus {
        val userId = requireUserId()
        return client.from("profiles")
            .select(Columns.list("dietary_setup_completed_at")) { filter { eq("id", userId) } }
            .decodeSingle()
    }

    // MARK: - Requirements

    /** Mirrors iOS `missingProfileRequirements`. */
    fun missingProfileRequirements(profile: Profile): List<ProfileRequirement> =
        buildList {
            if (profile.firstName?.trim().isNullOrEmpty()) add(ProfileRequirement.FIRST_NAME)
            if (profile.lastName?.trim().isNullOrEmpty()) add(ProfileRequirement.LAST_NAME)
            if (profile.username?.trim().isNullOrEmpty()) add(ProfileRequirement.USERNAME)
            if (profile.birthDate?.trim().isNullOrEmpty()) add(ProfileRequirement.BIRTH_DATE)
        }

    /** Mirrors iOS `date(from:)` — parses the stored `yyyy-MM-dd` birth date. */
    fun birthDate(from: String?): LocalDate? =
        from?.takeIf { it.isNotBlank() }?.let { runCatching { LocalDate.parse(it) }.getOrNull() }
}

// Update payloads. Kept private to this file, matching iOS.

/**
 * iOS declares `InitialProfileUpdate` and `ProfileUpdate` with identical
 * fields and coding keys; one type covers both here.
 */
@Serializable
private data class ProfileUpdate(
    @SerialName("first_name") val firstName: String,
    @SerialName("last_name") val lastName: String,
    @SerialName("display_name") val displayName: String,
    val username: String,
    val location: String,
    @SerialName("birth_date") val birthDate: String,
    val bio: String,
)

@Serializable
private data class ProfilePhotoUpdate(
    @SerialName("profile_photo_path") val profilePhotoPath: String,
)

@Serializable
private data class DietarySetupStatus(
    @SerialName("dietary_setup_completed_at") val dietarySetupCompletedAt: Instant? = null,
)

@Serializable
private data class DietarySetupReviewUpdate(
    @SerialName("dietary_setup_completed_at") val dietarySetupCompletedAt: Instant,
)
